pkg/loop/internal/median.go

```diff
@@ -42,39 +43,53 @@ func NewPluginMedianClient(broker Broker, brokerCfg BrokerConfig, conn *grpc.Cli
 	return &PluginMedianClient{pluginClient: pc, median: pb.NewPluginMedianClient(pc), serviceClient: newServiceClient(pc.brokerExt, pc)}
 }
 
+// TODO move; rename
+type HasClientConn interface {
+	// ClientConn returns the underlying connection to support proxying.
+	ClientConn() grpc.ClientConnInterface
+}
+
 func (m *PluginMedianClient) NewMedianFactory(ctx context.Context, provider types.MedianProvider, dataSource, juelsPerFeeCoin median.DataSource, errorLog ErrorLog) (ReportingPluginFactory, error) {
 	cc := m.newClientConn("MedianPluginFactory", func(ctx context.Context) (id uint32, deps resources, err error) {
-		dataSourceID, dsRes, err := m.serve("DataSource", func(s *grpc.Server) {
+		dataSourceID, dsRes, err := m.serveNew("DataSource", func(s *grpc.Server) {
 			pb.RegisterDataSourceServer(s, &dataSourceServer{impl: dataSource})
 		})
 		if err != nil {
 			return 0, nil, err
 		}
 		deps.Add(dsRes)
 
-		juelsPerFeeCoinDataSourceID, juelsPerFeeCoinDataSourceRes, err := m.serve("JuelsPerFeeCoinDataSource", func(s *grpc.Server) {
+		juelsPerFeeCoinDataSourceID, juelsPerFeeCoinDataSourceRes, err := m.serveNew("JuelsPerFeeCoinDataSource", func(s *grpc.Server) {
 			pb.RegisterDataSourceServer(s, &dataSourceServer{impl: juelsPerFeeCoin})
 		})
 		if err != nil {
 			return 0, nil, err
 		}
 		deps.Add(juelsPerFeeCoinDataSourceRes)
 
-		providerID, providerRes, err := m.serve("MedianProvider", func(s *grpc.Server) {
-			pb.RegisterServiceServer(s, &serviceServer{srv: provider})
-			pb.RegisterOffchainConfigDigesterServer(s, &offchainConfigDigesterServer{impl: provider.OffchainConfigDigester()})
-			pb.RegisterContractConfigTrackerServer(s, &contractConfigTrackerServer{impl: provider.ContractConfigTracker()})
-			pb.RegisterContractTransmitterServer(s, &contractTransmitterServer{impl: provider.ContractTransmitter()})
-			pb.RegisterReportCodecServer(s, &reportCodecServer{impl: provider.ReportCodec()})
-			pb.RegisterMedianContractServer(s, &medianContractServer{impl: provider.MedianContract()})
-			pb.RegisterOnchainConfigCodecServer(s, &onchainConfigCodecServer{impl: provider.OnchainConfigCodec()})
-		})
+		var (
+			providerID  uint32
+			providerRes resource
+		)
+		if grpcProvider, ok := provider.(HasClientConn); ok { //TODO ensure both paths are tested
+			providerID, providerRes, err = m.serve("MedianProvider", proxy.NewProxy(grpcProvider.ClientConn()))
+		} else {
+			providerID, providerRes, err = m.serveNew("MedianProvider", func(s *grpc.Server) {
+				pb.RegisterServiceServer(s, &serviceServer{srv: provider})
```

Comment:
```suggestion
		if grpcProvider, ok := provider.(HasClientConn); ok { //TODO ensure both paths are tested
		        //Provider supports proxying. Communication path is plugin server <-> proxy <-> plugin client
			providerID, providerRes, err = m.serve("MedianProvider", proxy.NewProxy(grpcProvider.ClientConn()))
		} else {
		        // The provider does not support proxying. Register services with the GRPC server directly.
                        // In this case, the communication path is plugin server <-> core client <-> core server <-> plugin client
			providerID, providerRes, err = m.serveNew("MedianProvider", func(s *grpc.Server) {
				pb.RegisterServiceServer(s, &serviceServer{srv: provider})
```

Am I understanding correctly that the branch in logic here chooses the communication path?

Reply:
Yes but the alternate path doesn't run quite like that, since it is only activated when the relay is running locally in legacy mode.

Reply:
I updated the README with a mermaid sequence diagram - lmk if that helps

Reply:
Sweet, the diagram definitely helps.

For line 74, In what case does `NewMedianFactory` receive a provider that needs to be wrapped with a grpc proxy vs not?

Reply:
When the relayer is running in a LOOP Plugin (relayer and product LOOPPs can be enabled/disabled independently).

Reply:
Ah - thanks!